distributed/scheduler.py

```diff
@@ -5648,143 +5650,172 @@ async def gather(self, keys, serializers=None):
+
+        if non_nannies:
+            raise ValueError(
+                (
+                    "Expected all workers to have a nanny, encountered "
+                    f"{len(non_nannies)} worker(s) without a nanny."
+                ),
+                non_nannies,
+            )
+
+    @log_errors
+    async def restart_workers(self, workers: Iterable[str], timeout: float = 30):
+        """
+        Restart a specified set of workers
+
+        This methods expects all workers to have nannies to be able to restart them.
+        If workers without nannies exist, ``Client.restart_workers`` will return a
+        serialized ``ValueError`` that lists the workers without nannies. Consider
+        removing those workers and calling ``Client.restart_workers`` again afterward.
+
+        Parameters
+        ----------
+        workers:
+            Workers to restart
+        timeout:
+            Return a `RuntimeError` if ``restart_workers`` takes more than ``timeout``
+            seconds
 
-        logger.debug("Send kill signal to nannies: %s", nanny_workers)
+        See Also
+        --------
+        Scheduler.restart
+        Client.restart_workers
+        """
+        try:
+            self._expect_nannies(workers)
+            logger.debug("Asking nannies to restart workers: %s.", workers)
+            await self._restart_workers(workers, timeout)
+            return {"status": "OK"}
+        except RuntimeError as e:
+            logger.exception(e)
+            return error_message(e)
+
+    async def _restart_workers(self, workers: Iterable[str], timeout: float) -> None:
+        nanny_addresses = [self.workers[worker].nanny for worker in workers]
         async with contextlib.AsyncExitStack() as stack:
             nannies = await asyncio.gather(
                 *(
                     stack.enter_async_context(
-                        rpc(nanny_address, connection_args=self.connection_args)
+                        rpc(address, connection_args=self.connection_args)
                     )
-                    for nanny_address in nanny_workers.values()
+                    for address in nanny_addresses
                 )
             )
 
-            start = monotonic()
-            resps = await asyncio.gather(
+            responses = await asyncio.gather(
                 *(
                     asyncio.wait_for(
-                        # FIXME does not raise if the process fails to shut down,
-                        # see https://github.com/dask/distributed/pull/6427/files#r894917424
-                        # NOTE: Nanny will automatically restart worker process when it's killed
-                        nanny.kill(
-                            reason="scheduler-restart",
-                            timeout=timeout,
-                        ),
-                        timeout,
+                        nanny.restart(timeout=timeout, reason="scheduler-restart"),
```

Comment:
Why are we calling `restart` here instead of `kill`? I don't trust `restart`.

1. `restart` doesn't pass the timeout into `kill`, so even if you have a 60s timeout here, `kill` will still fail if the worker doesn't shut down in 2s
2. `restart` manually calls `instantiate`, which is redundant because when the worker process exits, it'll call [`_on_worker_exit`](https://github.com/dask/distributed/blob/1e078f4afddad96e087cc6f8c23a40d4ae595017/distributed/nanny.py#L520-L552) which also calls `instantiate`. `instantiate` _should_ be idempotent so it _should_ be okay, but it feels like weird design. There should be exactly one thing responsible for restarting the worker process.

I assume the reason for this is to make the RPC block until the worker has connected to the scheduler, instead of having to do the wait-for-workers loop scheduler-side. That's reasonable, but only if we fix `Nanny.restart` first I think.

Reply:
Design-wise, I also don't love
1. `restart` being the Nanny's responsibility
2. the Nanny being responsible for deciding (or even knowing) when the worker has reconnected to the scheduler

The Nanny currently has a very complicated approach to subprocess management, and a lot of confusing IPC with queues, translating between blocking events and asyncio, etc.

I would love if the Nanny could (in theory) be rewritten in Go or even a bash script as a simple dumb process manager that just does its one job. If the process shuts down with a non-zero exit code, it restarts it. That's it. Maybe it sets some environment variables in the subprocess. Maybe it even still supports nanny plugins to run custom Python code. (Though I wish it couldn't even talk to the scheduler.) But I would love for the worker process to be a black box to the Nanny. They certainly shouldn't need to communicate or share any state; nothing beyond POSIX signals and an exit code.

If that were the case, then `Scheduler.restart` could be implemented as:
1. tell workers to close with non-zero exit code
2. wait for workers to reconnect because Nanny restarts them

Or, if the Nanny can communicate with the scheduler (maybe handy when a worker is busted and unresponsive), then `Scheduler.restart` could be:
1. call `Nanny.restart` RPC. This sends SIGTERM to the subprocess, joins with timeout, if it's not dead yet sends SIGKILL, joins, etc. Once the subprocess is dead, starts a new subprocess and returns.
2. wait for workers to reconnect because Nanny restarts them

The point is, I think we want `Scheduler.restart` to block until all workers have reconnected to the scheduler. _And I don't think the Nanny should be able to know if a subprocess is connected to the scheduler or not_. Therefore, that blocking logic should happen scheduler-side.

Reply:
> Why are we calling restart here instead of kill? I don't trust restart.

Just want to point out that the entire argument in the first comment is about the current implementation of the two handlers. If the handlers don't work as expected, we should fix them and not avoid using them. 

> restart being the Nanny's responsibility

IMO the nanny is the only sane place to implement this. The queue stuff is unnecessarily complex but that doesn't change anything about the way responsibilities should be assigned. From an ownership perspective, the nanny owns the worker resources and is responsible for closing and restarting it.

> I would love if the Nanny could (in theory) be rewritten in Go or even a bash script as a simple dumb process manager that just does its one job. If the process shuts down with a non-zero exit code, it restarts it. That's it. Maybe it sets some environment variables in the subprocess. Maybe it even still supports nanny plugins to run custom Python code. (Though I wish it couldn't even talk to the scheduler.) But I would love for the worker process to be a black box to the Nanny. They certainly shouldn't need to communicate or share any state; nothing beyond POSIX signals and an exit code.

What problem is this actually solving? We'd abandon queues in favor of bash?? How would Go help? These are all implementation details but if we don't fix the fundamental architecture, we'll still be in the same mess.

Reply:
FWIW `Nanny.kill` currently suffers from a race condition `Nanny.restart` apparently doesn't https://github.com/dask/distributed/issues/7312#issuecomment-1315316191

In that issue I intend to clean this logic up a bit anyhow which should resolve the tension here a bit, I believe.

Reply:
> Why are we calling `restart` here instead of `kill`? I don't trust `restart`.

`restart` is imperfect, we agree on that. Fixing it is something we should do, but I would consider this outside the scope of this PR. The scope of this PR is to simplify the scheduler-side of restarting, i.e., `Scheduler.restart`, remove a number of implicit assumptions and raise exceptions in their place, and ensure that `Client.restart` and `Client.restart_workers` actually have similar semantics.

I agree with Florian that restarts should be the responsibility of the nanny. Arguably, that's all the nanny has to do: Watch over the worker process and restart it when told to. We also need a way to be able to kill/restart a struggling/unresponsive worker, which would not be possible to implement on the worker since it's, well, unresponsive.